### DISCUSSION TITLE
How to deal with missing glpyhs in fonts

### DISCUSSION THREAD

**Discussion by milasudril** [harfbuzz/harfbuzz#4949] [Q&A] [answered]
Suppose I load a font and render some text, but a glyph is missing in the selected font so I get a square after rendering. What API:s should I use to provide a list of alternate fonts?

**Comment by khaledhosny** [comment]
HarfBuzz does not handle font fallback (or any font change for that matter), buffers are always shaped with a single font. Font fallback is handled by the client.

**Reply by milasudril** [reply]
I know that, I am more asking about detection, and what actions to take when there is a missing block if missing glyphs.

**Reply by khaledhosny** [reply, marked as answer]
There are usually two strategies to handle font fallback:
1. Check font character coverage before shaping and for any sequence of characters the current font doesn't support try the next font on the list.
2. Shape, then for as any sequence of unsupported characters, shape with the next font on the list.

**Reply by behdad** [reply]
> 2\. Shape, then for as any sequence of unsupported characters, shape with the next font on the list.

You should do the fallback per Unicode Grapheme cluster, or if you don't have code for that, per HarfBuzz cluster, for each cluster that has unshaped glyphs. Those are, by default, glyph index 0. You can change it using HB API if you want for any reason, but that typically doesn't apply.
